README.md
# COOP

[![npm](https://img.shields.io/npm/v/chickencoop?style=for-the-badge)](https://www.npmjs.com/package/chickencoop)

Controller Operational Observation Program

---
COOP is a JavaScript program that lets you generate a button binding reference sheet for Xbox controllers.

## Usage

```
npx chickencoop -i /path/to/input -o /path/to/output

Options:
      --help     Show help                                             [boolean]
  -i, --input    The file to read from                       [string] [required]
  -o, --output   The file to write to                        [string] [required]
```

## Syntax

COOP will look for special lines in the input file. If your input file is a source code file, these lines can be added as comments.

```
coop:button(buttonName,text,controller)
```

### Button Name

This is the button's identifier, can be any of the following:

- `LJoystick`
- `RJoystick`

- `LBumper`
- `RBumper`

- `LTrigger`
- `RTrigger`

- `DPadUp`
- `DPadDown`
- `DPadLeft`
- `DPadRight`

- `A`
- `B`
- `X`
- `Y`

### Text

This will be the label text for the button. It can have spaces, but no commas or newlines

### Controller

Can be either `pilot` or `copilot`. `pilot` is the top controller in the diagram and `copilot` is the bottom one.

## Example

See the [example input](example.txt) and [its output](example.pdf)


CHANGELOG.md
# Changelog

### v2.1.0

- Add start and back buttons

## v2.0.0

- Rewritten in TypeScript
- Lines have adjustable lengths, depending on labels width
- Labels have a maximum width
- Will not allow you to set multiple labels on the same button


script.js
#!/usr/bin/env node

const { createCanvas, loadImage } = require("canvas");
const fs = require("fs");
const canvas = createCanvas(780, 800, "pdf");
const ctx = canvas.getContext("2d");

const yargs = require('yargs');

const argv = yargs
                .scriptName("npx chickencoop")
                .usage('$0 -i /path/to/input -o /path/to/output.pdf')
                .option('input', {
                    alias : 'i',
                    describe: 'The file to read from',
                    type: 'string', /* array | boolean | string */
                    nargs: 1,
                    demand: true
                })
                .option('output', {
                    alias : 'o',
                    describe: 'The file to write to',
                    type: 'string', /* array | boolean | string */
                    nargs: 1,
                    demand: true
                })
                .example("$0 -i input.txt -o output.pdf")
                .showHelpOnFail(false, "Specify --help for available options")

                .argv


locations = {
	//"Location": [[[X, Y]...], "textAlign", "textColor", [labelStartOffsetX, labelStartOffsetY]],

	LJoystick: [
		[
			[80, 115],
			[80, 120],
			[280, 120],
		],
		"left",
		"black",
		[-70, -10],
	],
	RJoystick: [
		[
			[700, 225],
			[700, 230],
			[500, 230],
			[463, 195],
		],
		"right",
		"black",
		[60, -10],
	],

	LBumper: [
		[
			[360, 35],
			[360, 40],
			[300, 61],
		],
		"right",
		"black",
		[20, -10],
	],
	RBumper: [
		[
			[440, 35],
			[440, 40],
			[500, 64],
		],
		"left",
		"black",
		[-20, -10],
	],

	LTrigger: [
		[
			[60, 55],
			[60, 60],
			[275, 63],
		],
		"left",
		"black",
		[-50, -10],
	],
	RTrigger: [
		[
			[700, 60],
			[520, 64],
		],
		"right",
		"black",
		[50, -10],
	],

	DPadUp: [
		[
			[80, 165],
			[80, 170],
			[340, 170],
		],
		"left",
		"black",
		[-70, -10],
	],
	DPadDown: [
		[
			[80, 270],
			[80, 275],
			[290, 275],
			[340, 220],
		],
		"left",
		"black",
		[-70, -10],
	],

	DPadLeft: [
		[
			[80, 200],
			[80, 205],
			[300, 205],
			[315, 195],
		],
		"left",
		"black",
		[-70, -10],
	],
	DPadRight: [
		[
			[80, 235],
			[80, 240],
			[290, 240],
			[365, 195],
		],
		"left",
		"black",
		[-70, -10],
	],

	A: [
		[
			[680, 155],
			[680, 160],
			[538, 160],
		],
		"right",
		"#7b8e4a",
		[70, -5],
	],
	B: [
		[
			[680, 125],
			[680, 130],
			[568, 130],
		],
		"right",
		"#ca452f",
		[70, -5],
	],

	Y: [
		[
			[680, 95],
			[680, 100],
			[538, 100],
		],
		"right",
		"#de9f08",
		[70, -5],
	],
	X: [
		[
			[680, 185],
			[680, 190],
			[500, 190],
			[495, 140],
		],
		"right",
		"#5d72c0",
		[70, -5],
	],
};

function drawLines(pointList, pilot) {
	let lineStartX = pointList[0][0];
	var lineStartY = getYCoord(pointList[0][1], pilot);

	ctx.beginPath();
	ctx.moveTo(lineStartX, lineStartY);

	for (var i = 1; i < pointList.length; i++) {
		ctx.lineTo(pointList[i][0], getYCoord(pointList[i][1], pilot));
	}

	ctx.stroke();
}

function getYCoord(point, pilot) {
	if (pilot) {
		return point;
	} else {
		return point + 400;
	}
}

function addButton(locationName, text, pilot) {
	var location = locations[locationName];

	drawLines(location[0], pilot);

	let textX = location[0][0][0] + location[3][0];
	let textY = getYCoord(location[0][0][1], pilot) + location[3][1];

	ctx.textAlign = location[1];
	ctx.fillStyle = location[2];
	ctx.fillText(locationName + ": " + text, textX, textY);
}

function initCanvas() {
	loadImage(require('path').join(__dirname,"background.png")).then((image) => {
		ctx.drawImage(image, 10, 10, 780, 800);
		ctx.font = "15px Arial";
        let file;
        try {
            file = fs.readFileSync(argv.input).toString();
        } catch {
            console.log("Path does not exist.")
            return
        }

		const matches = file.matchAll(
			/coop:button\((LJoystick|LBumper|RJoystick|RBumper|LTrigger|RTrigger|DPadUp|DPadDown|DPadLeft|DPadRight|B|A|X|Y),([^,\n]+),(pilot|copilot)\)/g
		);
		for (let match of matches) {
			addButton(match[1], match[2], (match[3]=="pilot"));
		}
		exportPDF();
	});
}


function exportPDF() {
	fs.writeFileSync(argv.output, canvas.toBuffer());
}

initCanvas();

src/index.ts
#!/usr/bin/env node


const fs = require("fs");

const { createCanvas, loadImage } = require("canvas");
const canvas = createCanvas(780, 800, "pdf");
const ctx:CanvasRenderingContext2D = canvas.getContext("2d");

const colors = require('colors/safe');

const yargs = require('yargs');

const argv = yargs
                .scriptName("npx chickencoop")
                .usage('$0 -i /path/to/input -o /path/to/output.pdf')
                .option('input', {
                    alias : 'i',
                    describe: 'The file to read from',
                    type: 'string', /* array | boolean | string */
                    nargs: 1,
                    demand: true
                })
                .option('output', {
                    alias : 'o',
                    describe: 'The file to write to',
                    type: 'string', /* array | boolean | string */
                    nargs: 1,
                    demand: true
                })
                .example("$0 -i input.txt -o output.pdf")
                .showHelpOnFail(false, "Specify --help for available options")
                .argv

class Point {
	x: number
	y: number
	private copilotY:number;
	
	constructor(x:number, y:number) {
		this.x = x
		this.y = y
		this.copilotY = y+400;
	}

	static fromArray(coordinates: number[]):Point {
		return new Point(coordinates[0], coordinates[1])
	}


	public getY(pilot:boolean):number {
		if (pilot) {
			return this.y
		} else {
			return this.copilotY
		}
	}
}
class Button {
	name: string
	points: Point[] = []
	text: {
		alignment: CanvasTextAlign,
		color: string
		anchor: Point
		maxWidth:number
	}
	private lock={
		pilot: false,
		copilot: false
	}

	constructor(name:string, points:number[][], textAlignment:CanvasTextAlign, textColor:string, maxWidth:number, textAnchor:[number, number]) {
		this.name = name
		points.forEach((point) => {
			this.points.push(Point.fromArray(point))
		})
		this.text = {
			alignment: textAlignment,
			color: textColor,
			anchor: Point.fromArray(textAnchor),
			maxWidth: maxWidth
		}
	}


	private checkLocks(pilot: boolean) {
		if (pilot) {
			if (this.lock.pilot) {
				console.warn(colors.yellow(`Binding for ${this.name} on Pilot has been registered twice!`))
				return true
			}
			this.lock.pilot = true
		} else {
			if (this.lock.copilot) {
				console.warn(colors.yellow(`Binding for ${this.name} on Copilot has been registered twice!`))
				return true
			} else {
				this.lock.copilot = true
			}
			
		}
		return false
	}
	public draw(label:string, pilot:boolean):void {
		if (!this.checkLocks(pilot)) {
			let lineEnd = this.drawText(label, pilot);
			this.drawLines(lineEnd, pilot)
		}
		
	}

	private drawText(label:string, pilot:boolean):Point {
		let text = this.name + ": " + label;
		let textX = this.text.anchor.x
		let textY =  this.text.anchor.getY(pilot)
		let maxWidth = this.text.maxWidth

		ctx.textAlign = this.text.alignment
		ctx.fillStyle = this.text.color

		ctx.fillText(text, textX,textY, maxWidth);


		let measuredWidth = Math.min(ctx.measureText(text).width, maxWidth)
		if (this.text.alignment == "left") {
			return new Point(textX+measuredWidth+5, textY)
		} else {
			return new Point(textX-measuredWidth-5, textY)
		}
	}

	private drawLines(labelLocation:Point, pilot:boolean) {
		ctx.beginPath();
		ctx.moveTo(labelLocation.x, labelLocation.y);

		this.points.forEach(point => {
			ctx.lineTo(point.x, point.getY(pilot));
		})
		ctx.stroke();
	}

}

let locations = {
	LJoystick: new Button("LJoy", [[280, 120]], "left", "black", 230, [10, 120]),
	RJoystick: new Button("RJoy", [[500, 230],[463, 195]], "right", "black", 250, [770,230]),

	LBumper: new Button("LB", [[270, 65]], "left", "black", 240, [10, 65]),
	RBumper: new Button("RB", [[530, 65],[530, 67]], "right", "black", 230, [770, 65]),

	LTrigger: new Button("LT", [[280, 40], [280, 62]], "left", "black", 250, [10,40]),
	RTrigger: new Button("RT", [[520, 40], [520, 65]], "right", "black", 240, [770,40]),

	A: new Button("A", [[539, 160]], "right", "#7b8e4a", 190, [770, 160]),
	B: new Button("B", [[568, 130]], "right", "#ca452f", 190, [770, 130]),
	Y: new Button("Y", [[538, 100]], "right", "#de9f08", 190, [770, 100]),
	X: new Button("X", [[500, 190],[495, 140]], "right", "#5d72c0", 190, [770, 190]),

	DPadUp:   new Button("DPad Up",   [[340,170]], "left", "black", 270, [10,170]),
	DPadDown: new Button("DPad Down", [[340,220]], "left", "black", 270, [10,220]),
	DPadLeft: new Button("DPad Left", [[315, 195]], "left", "black", 270, [10,195]),
	DPadRight: new Button("DPad Right", [[365, 245],[365, 195]], "left", "black", 270, [10,245]),

	Back: new Button("Back", [[360, 15],[360, 122]], "left", "black", 330, [10, 15]),
	Start: new Button("Start", [[447, 15],[447, 122]], "right", "black", 300, [770, 15]),
};

class CoopMatch {
	button: Button
	buttonName: string
	label: string
	pilot: boolean
	constructor(annotation:RegExpMatchArray) {
		this.buttonName = annotation[1]
		this.button = locations[this.buttonName]
		this.label = annotation[2]
		this.pilot = (annotation[3].toString() == "pilot")
	}

	public draw():void {
		if (!this.button) {
			console.error(colors.yellow("Could not find button '%s'"), this.buttonName)
		} else {
			this.button.draw(this.label, this.pilot)
		}
	}
}

function initCanvas() {
	loadImage(require('path').join(__dirname,"assets/background.png")).then((image) => {
		ctx.drawImage(image, 10, 10, 780, 800);
		ctx.font = "15px Arial";
		ctx.textBaseline = "middle"
        let file:string;
        try {
            file = fs.readFileSync(argv.input).toString();
        } catch {
            console.error(colors.red("Path does not exist."))
			process.exit(1)
        }
		const matches = file.matchAll(
			/coop:button\(([^,\n]+),([^,\n]+),(pilot|copilot)\)/g
		);
		for (let match of matches) {
			new CoopMatch(match).draw()
		}
		exportPDF();
	});
}


function exportPDF() {
	fs.writeFileSync(argv.output, canvas.toBuffer());
}

initCanvas();